text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+    // `(k, maps)` declares the iterated variadic tuples `k` and `maps`
+    // `<K, V>` declares the iterated variadic tuple types
+    (for (ref key, map) <KEY, VALUE> @in (k, maps) <K, V> {
+        HashMap::<KEY, VALUE>::get(&map, key)
+    })
+};
+```
+
+Note: when iterating over multiple variadic tuple or variadic tuple types, they must have all the same arity. To ensure this, all variadic tuple types involved must have been declared together.
+
+Examples:
+
+```rust
+impl<(..(K, V))> MegaMap<(..(K, V))>
+where ..(K: Hash), {
+    fn get(&self, k: (..K)) -> (..Option<V>) {
+        let result: (..Option<&V>) = {
+            (for (ref k, map) <Key, Value> @in (k, &self.maps) <K, V> {
+                HashMap::<Key, Value>::get(&map, k)
+            })
+        };
+
+        result
+    }
+}
+
+impl<(..T), Last> Hash for (..T, Last)
+where
+    ..(T: Hash),
+    Last: Hash + ?Sized, {
+
+    #[allow(non_snake_case)]
+    fn hash<S: Hasher>(&self, state: &mut S) {
+        let (ref tuple @ .., ref last) = *self;
+       
+        // Use case: only variadic tuple
+        for member @in tuple {
+          member.hash(state);
+        };
+        last.hash(state);
+
+        // Use case: variadic tuple and type
+        for member <H> @in tuple <T> {
+          <T as Hash>::hash(&member, state);
+        };
+        last.hash(state);
+    }
+}
+
+trait Merge<(..R)> {
+    type Value;
+    fn merge(self, r: (..R)) -> Self::Value;
+}
+
+impl<(..L), (..R)> Merge<(..R)> for (..L) {
+    type Value = (..L, ..R);
+
+    fn merge(self, r: (..R)) -> Self::Value {
+        (
+            for l1 @in self { l1 },
+            for r1 @in r { r1 },
+        )
+    }
+}
+
+trait Integer {
+    fn one() -> Self;
+}
+
+fn add_one<(..T)>((..t): (..T)) -> (..T)
+where
+    ..(T: Integer + Add), {
+    (for t1 <T1> @in t <T> { t1 + T1::one() })
+}
+```
+
+## The `Hash` trait
+
+Let's implement the `Hash` trait:
+
+```rust
+// For the example, we consider the impl for (A, B, C). So `(..T)` matches `(A, B, C)`
+// We have the first expansion here, `(..T, Last)` expands to `(A, B, C, Last)`
+impl<(..T), Last> Hash for (..T, Last) 
+where
+        // Expands to `A: Hash, B: Hash, C: Hash,`
+    ..(T: Hash,),
+    Last: Hash + ?Sized, {
+
+    #[allow(non_snake_case)]
+    fn hash<S: Hasher>(&self, state: &mut S) {
+        // Destructure self to a variadic tuple `tuple` and a variable `last`. The variadic tuple type of `tuple` is `(..&T)`
+        // So it will be equivalent to `let (ref a, ref b, ref c, ref last) = *self; let tuple = (a, b, c);`
+        let (ref tuple @ .., ref last) = *self;
+        for member @in tuple {
+          member.hash(state);
+        };
+        // The for loop will be inlined as: 
+        // ( 
+        //    { v.0.hash(state); },
+        //    { v.1.hash(state); },
+        //    { v.2.hash(state); },
+        // );
+        last.hash(state);
+    }
+}
+```
```

Comment:
Like I mentioned earlier, an example like this should really be put throughout the section, rather than at the end. `Hash` is a good example though, and I'd recommend using it.